### ISSUE TITLE
Lack of code-signing makes installation hard on Windows and Mac

### ISSUE THREAD

**Issue by petebankhead** [qupath/qupath#1731]
This comes up relatively often, e.g.

* https://github.com/qupath/qupath/issues/1672
* https://github.com/qupath/qupath/issues/1605

It's not something we have the time, resources or developer accounts to deal with right now, but this issue exists as a single reference for the future.

**Comment by rougegoat** [comment]
Lack of signing now prevents the official pkg from being deployed.  Gatekeeper prevents it from running.

**Comment by petebankhead** [comment]
Unless there has been a **very** recent update that breaks it, the instructions at https://qupath.readthedocs.io/en/stable/docs/intro/installation.html#macos should still work.

**Comment by rougegoat** [comment]
You shouldn't need to work around basic security settings built into macOS to install a program.

There's also a big issue with a managed deployment caused by the lack of signing.  In situations where the end user is not an admin (incredibly common in education and commercial), they will not be able to manually allowlist the application every time an update comes out.

**Comment by uurazzle** [comment]
Hi:

Thanks for maintaining and making QuPath available for macOS. 

I understand that code signing and notarization require ongoing cost, infrastructure, and maintenance effort, and that this may not be something the project currently has resources to support. However, I wanted to emphasize that the impact has increased significantly for managed macOS deployments in education, research, and enterprise environments.

Recent versions of macOS rely heavily on Gatekeeper, Developer ID signing, and notarization for software distributed outside the App Store. When applications or installer packages are unsigned or unnotarized, administrators often cannot deploy them through standard management workflows without additional exceptions or user intervention. In many organizations, end users do not have administrative rights and therefore cannot approve each new release manually. Apple's documentation describes notarization as the recommended path for software distributed outside the Mac App Store and notes that Gatekeeper uses notarization tickets to verify software before launch.

For future consideration, the relevant Apple documentation includes:

 - Apple: Notarizing macOS software before distribution
https://developer.apple.com/documentation/security/notarizing-macos-software-before-distribution

 - Apple: Code Signing Services
https://developer.apple.com/documentation/security/code-signing-services

Even if full implementation is not currently feasible, it may be useful to document code signing and notarization as a longer-term project goal because it would substantially improve deployment in managed environments and reduce the need for users to bypass built-in platform security controls. The current workaround may function for individual users, but it remains a significant barrier for institutional deployments.

In addition, the lack of code signing and notarization limits the ability to use widely adopted macOS automation tools such as [AutoPkg](https://github.com/autopkg/autopkg) and [Installomator](https://github.com/Installomator/Installomator). Many organizations rely on these tools to automate software packaging, deployment, and updates. Without a trusted signing and notarization chain, automated deployment workflows become significantly more complex or impossible to implement within standard security policies, creating an additional barrier to adoption in managed environments.

**Comment by petebankhead** [comment]
We're aware it's a problem. The existence of this GitHub issue is our way of documenting our desire to overcome it.

Nevertheless, I don't expect the situation will change any time soon. QuPath is open-source software developed by a very small research group. There is no commercial entity behind it. We are not selling anything or trying to maximise users. We [have made it as easy to build QuPath as we can](https://qupath.readthedocs.io/en/stable/docs/reference/building.html) and produce binaries as a convenience, but our focus is on the research side.

If someone needs to create their own signed builds for deployment, they can try [editing this file](https://github.com/qupath/qupath/blob/main/buildSrc/src/main/kotlin/qupath.jpackage-conventions.gradle.kts) and checking out the [jpackage docs](https://docs.oracle.com/en/java/javase/26/docs/specs/man/jpackage.html). I'm aware that another developer once submitted a version of QuPath for notarisation with their own developer ID to check if it was possible. I understand that it worked.

For us, the issues are not *only* ongoing cost, infrastructure and maintenance effort. We also need to think about whether there are any other implications or risks around any individual using their own developer ID, and considering things like license compatibility, dependencies, and support for all relevant platforms. There might not be any major problems, but it would take work (and reading a lot of dull documents) for me to feel confident in it. None of this was really an issue when QuPath was first released a decade ago and the systems that are available seem better designed for companies with legal teams (and revenue).

So while we certainly don't want QuPath users to be inconvenienced, we are not individually willing or able to address all the inconveniences at this time. We don't do everything users want, but I think we do a lot (mostly unseen).

Over the next year or two, we want to move towards the core QuPath application being much smaller and more streamlined. Figuring out how to distribute the software will play a big role in this redesign, since the current approach was fixed before these issues arose.

**Comment by alanocallaghan** [comment]
To chip in a bit from the perspective of a Linux user (and QuPath developer) I agree that users ought not to need to circumvent security settings in order to install software. However, placing all of the onus on open source developers, rather than the massive companies (two of the largest in the world!) that develop these increasingly locked-down operating systems seems a touch unfair.

We are aware of the issue, and if it were a simple and quick fix then we would have deployed it many years ago at this point. Thanks for raising these suggestions.

**Comment by uurazzle** [comment]
I completely understand the concern about time, resources, and the overhead of maintaining developer accounts and signing infrastructure.

That said, it seems that a growing number of open-source projects have addressed this by treating signing and notarization as part of the release pipeline rather than part of the development process.

### Examples

* **DDEV** maintains reusable open-source tooling for macOS notarization and Windows signing that runs entirely in CI. Their approach is specifically designed to reduce the burden on maintainers by automating the release process rather than requiring manual signing steps.
  * https://github.com/ddev/signing_tools

* Many desktop application projects built with **Electron** now use automated code signing and notarization during CI/CD releases.
  * Electron Code Signing Guide: https://www.electronjs.org/docs/latest/tutorial/code-signing
  * Electron Builder Code Signing: https://www.electron.build/code-signing

* **Tauri** includes built-in support and documentation for automated code signing and notarization workflows.
  * https://tauri.app/distribute/sign/
  * https://tauri.app/distribute/sign/macos/

From the perspective of universities, research institutes, hospitals, and enterprise environments, unsigned applications increasingly create deployment and support challenges. Security controls such as Windows SmartScreen and macOS Gatekeeper often generate warnings that users either cannot bypass or are instructed not to bypass.

References:
* Apple Gatekeeper Overview:
  https://support.apple.com/guide/security/gatekeeper-and-runtime-protection-sec5599b66df/web
* Microsoft SmartScreen Overview:
  https://learn.microsoft.com/en-us/windows/security/operating-system-security/virus-and-threat-protection/microsoft-defender-smartscreen/

As a result, the lack of signing can become more than a usability issue—it can become a barrier to institutional adoption and managed deployment.

**Comment by alanocallaghan** [comment]
From

> https://github.com/ddev/signing_tools

Mentions

> Notarization requires
>    An Apple Developer Program organization membership from developer.apple.com

[Enroll page](https://developer.apple.com/programs/enroll/) requirements include

> Legal entity name and status. Your organization must be a legal entity that can enter into contracts with Apple.

As there is no QuPath LLC, this in our case would presumably be our university. Creating this account would therefore likely require consultation with the university's legal department at a minimum (perhaps also consulting with IT and some external bodies). I assume the fun would only multiply if we ever had to move to a different university.

Now I love killing time as much as the next person, but I'm sure my enthusiasm for a lengthy back and forth with legal about software signing is only surpassed by @petebankhead 's. In the meantime we have ongoing collaborations relating to QuPath, method development, software maintenance, user help requests/forum posts, bug fixes, refactoring, [...], which are usually given priority.

**Comment by janxkoci** [comment]
As an alternative, would you consider distributing binaries via `conda` or `brew`?

I already see some version of [QuPath on conda](https://anaconda.org/channels/sdvillal/packages/qupath/overview), in a personal channel - is that in any way associated with this project? It offers binaries for all 3 platforms. There is one more [binary for linux](https://anaconda.org/channels/HCC/packages/qupath/overview) distributed by HCC, which I've at least heard of, they have lot of other packages in their channel.

In my line of work (genomics), `conda` is nowadays almost a standard, but I don't use QuPath myself - I'm looking into it for my wife, who has a macbook, and she'd run into this issue.

PS: perhaps it's worth adding, that most of the software I use at work is distributed by either [Bioconda](https://bioconda.github.io/) or [Conda-forge](https://conda-forge.org/) channels. As the name suggests, Bioconda focuses on bioinformatic software, and to my knowledge they distribute predominantly Linux and MacOS binaries, although they may have added some Windows binaries lately. (It's also possible to use Bioconda software through WSL.) Conda-forge on the other hand focuses on general scientific software, and often supports also Windows. So you could submit QuPath to these channels, or even create your own personal (public) channel and distribute it there. Also, GUI apps can be distributed too - I know Rstudio, IGV, and Tablet are available via conda, which provides dependencies like Xorg or Wayland (I only use Linux, but I imagine they provide some GUI stack for the other platforms too).

**Comment by petebankhead** [comment]
Yes, that's something we're considering. The existing packages have been created by others, not by the QuPath devs.

I'm not sure if it entirely solves the code signing issue though, and I'd guess that it's only a minority of QuPath's users that would already be using `conda` or `brew`. Quite a lot would likely find the installation experience worse, so we'd probably end up needing to maintain various options. We're somewhat overwhelmed by projects right now, so it's not something we're in a position to investigate at the moment.

**Comment by janxkoci** [comment]
Yes, I'm well aware that typing three words into a terminal can be too much for some. On the other hand the instructions posted above to workaround Apple security look quite intimidating as well. 

However, conda can solve several issues discussed in this thread:

- conda works without admin privileges - we use it at our university in some courses and it's great that we don't need to ask tech support to install anything. We can also show students how to get it on their own machines and it works all the same.
- the above also means your "unverified" software cannot break user's systems, so they may feel better about that 😉
- even if there was signing (which I'm not sure there is), I doubt it's anywhere near as involved as the Big Tech requires. Conda packages are regularly published by solo devs without legal teams, even into those big channels I've mentioned earlier.

Last but not least, there are some graphical interfaces to conda. The Anaconda corp provides their Navigator, but I've also seen a conda package manager plugin for the Spyder IDE (for scientific python coding), for example. 

But I understand there is lots of things to consider, I just wanted to throw it out there so people here know about these options. Take it or leave it. Or maybe someone else will pick it and run with it 😉

**Comment by fhalwani** [comment]
I don't use a Mac, but I bypassed all the security on Windows 11 PC at work by installing QuPath in my home directory. I created a subdir and unzipped the binary there. It worked like a charm, I've done that for a dozen other applications (GIMP, FiJi, dBeaver, Krita, vscode) and no issues, unless the workstation is for multiple users, then each user has to install a copy in their respective home folders. The benefits of this amazing open source software outweighs this small inconvenience.
